src/txn/txn.c

```diff
@@ -140,42 +140,43 @@ __wt_txn_release_snapshot(WT_SESSION_IMPL *session)
 
 /*
  * __wt_txn_user_active --
- *     Check whether there are any running user transactions. Note that a new transactions may start
- *     on a session we have already examined and the caller needs to be aware of this limitation.
- *     Exclude prepared user transactions from this check.
+ *     Check whether there are any running user transactions. Note that a new transaction may start
+ *     after we return from this call and therefore caller should be aware of this limitation.
  */
 bool
 __wt_txn_user_active(WT_SESSION_IMPL *session)
 {
     WT_CONNECTION_IMPL *conn;
     WT_SESSION_IMPL *session_in_list;
+    WT_TXN_GLOBAL *txn_global;
+    WT_TXN_SHARED *txn_shared;
     uint32_t i, session_cnt;
     bool txn_active;
 
     conn = S2C(session);
     txn_active = false;
+    txn_global = &conn->txn_global;
+
+    /* We're going to scan the table: wait for the lock. */
+    __wt_readlock(session, &txn_global->rwlock);
 
-    /*
-     * No lock is required because the session array is fixed size, but it may contain inactive
-     * entries. We must review any active session, so insert a read barrier after reading the active
-     * session count. That way, no matter what sessions come or go, we'll check the slots for all of
-     * the user sessions for active transactions when we started our check.
-     */
     WT_ORDERED_READ(session_cnt, conn->session_cnt);
-    for (i = 0, session_in_list = conn->sessions; i < session_cnt; i++, session_in_list++) {
-        /* Skip inactive sessions. */
-        if (!session_in_list->active)
+    WT_STAT_CONN_INCR(session, txn_walk_sessions);
+    for (i = 0, session_in_list = conn->sessions, txn_shared = txn_global->txn_shared_list;
+         i < session_cnt; i++, session_in_list++, txn_shared++) {
+
+        /* Skip inactive or internal sessions. */
+        if (!session_in_list->active || F_ISSET(session_in_list, WT_SESSION_INTERNAL))
             continue;
-        /* Check if a user session has a running transaction. Ignore prepared transactions. */
-        if (F_ISSET(session_in_list->txn, WT_TXN_RUNNING) &&
-          !F_ISSET(session_in_list, WT_SESSION_INTERNAL) &&
-          !F_ISSET(session_in_list->txn, WT_TXN_PREPARE)) {
 
+        /* Check if a user session has a running transaction. */
+        if (txn_shared->id != WT_TXN_NONE || txn_shared->pinned_id != WT_TXN_NONE) {
```

Comment:
Once a transaction rolls back, commits or get prepared, `__txn_remove_from_global_table()` is called that should clear both `id` and `pinned_id`. As we are holding global transaction lock, no new transaction can start while we are looping.